### PR TITLE
PERF: df.groupby(categorical, sort=False)

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

xref: #48749

```
In [1]: df = pd.DataFrame({"key": pd.Categorical(range(1_000_000 - 1, -1, -1), range(1_000_000))})

In [2]: %timeit df.groupby("key", sort=False)  # PR
66.7 ms ± 1.53 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [9]: %timeit df.groupby("key", sort=False)  # main
586 ms ± 4.15 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

### CODE REVIEW

File: pandas/core/groupby/categorical.py [hunk 1]
```diff
@@ -75,21 +75,22 @@ def recode_for_groupby(
...
-    # including those missing from the data (GH-13179), which .unique()
-    # above dropped
-    cat = cat.add_categories(c.categories[~c.categories.isin(cat.categories)])
-
-    return c.reorder_categories(cat.categories), None
+    # xref GH:46909: Re-ordering codes faster than using (set|add|reorder)_categories
+    all_codes = np.arange(c.categories.nunique(), dtype=np.int8)
+    # GH 38140: exclude nan from indexer for categories
+    # TODO: Probably why dropa=False doesn't work GH 36327
+    unique_notnan_codes = unique1d(c.codes[c.codes != -1])
```
Comment (rhshadrach) [inline pandas/core/groupby/categorical.py:83]: I'm working on a branch that fixes `dropna=False` for categorical (currently writing tests, but at least reducers and transformers seem to be working), this section of code is unchanged in that. My thinking here is that this is determining the categories and categories cannot include nan, so dropping -1 is okay even for `dropna=False`.

**Review by rhshadrach — APPROVED** [review]
lgtm, nice speedup!

File: pandas/core/groupby/categorical.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/groupby/categorical.py:83]: Ah okay thanks for the insights. I'll remove my comment then referencing dropna=False.

I just found it odd that the this generates `self.grouping_vector` and it specifically drops -1 codes unlike `sort=True`

File: pandas/core/groupby/categorical.py [hunk 2]
```diff
@@ -75,21 +75,21 @@ def recode_for_groupby(
...
-    cat = cat.set_categories(cat.categories.take(take_codes))
-
-    # But for groupby to work, all categories should be present,
-    # including those missing from the data (GH-13179), which .unique()
-    # above dropped
-    cat = cat.add_categories(c.categories[~c.categories.isin(cat.categories)])
-
-    return c.reorder_categories(cat.categories), None
+    # xref GH:46909: Re-ordering codes faster than using (set|add|reorder)_categories
+    all_codes = np.arange(c.categories.nunique(), dtype=np.int8)
```
Comment (rhshadrach) [inline pandas/core/groupby/categorical.py:80]: @mroeschke - why the specification of int8 here? This causes overflow when there are more than 256 categories. Found this via the failing asv in #49131 - if it's okay, going to remove the specification of dtype there.

Comment (mroeschke) [reply pandas/core/groupby/categorical.py:80]: I think I was trying to match the output of `Categorical.codes` here which was int8, but here specifically I think ensuring the `dtype` is `int` should be sufficient.

Comment (mroeschke) [reply pandas/core/groupby/categorical.py:80]: Feel free to change it in https://github.com/pandas-dev/pandas/pull/49131

**Pull request merged**